src/budget/NewCategoryPage.js

```diff
@@ -19,18 +22,34 @@ export default function NewCategoryPage() {
   const [name, setName] = useState('');
   const [target, setTarget] = useState('');
 
-  const handleCreateCategory = () => {
-    const newCategory = {
-      name,
-      target,
-    };
-    dispatch(addNewCategory(newCategory));
-    navigate('/budget');
-  };
+  const [errors, setErrors] = useState({ name: [], target: [] });
 
   const handleTargetChange = (event) => {
-    const targetInput = parseInt(event.target.value, 10);
-    setTarget(targetInput);
+    const targetInput = event.target.value;
+    setTarget(targetInput * 1);
+    setErrors((currentErrors) => ({ ...currentErrors, target: validate(targetInput, 'Target', [required, restrictedRange, isNumber]) }));
+  };
+
+  const handleNameChange = (event) => {
+    const nameInput = event.target.value;
+    setName(nameInput);
+    setErrors((currentErrors) => ({ ...currentErrors, name: validate(nameInput, 'Name', [required, certainCharactersOnly, max50Characters]) }));
+  };
+
+  // Creates a new category on form submission
+  const handleCreateCategory = () => {
+    const nameErrors = validate(name, 'Name', [required, certainCharactersOnly, max50Characters]);
+    const targetErrors = validate(target, 'Target', [required, restrictedRange, isNumber]);
+    setErrors((currentErrors) => ({ ...currentErrors, name: nameErrors }));
+    setErrors((currentErrors) => ({ ...currentErrors, target: targetErrors }));
+    if ((nameErrors.length === 0) && (targetErrors.length === 0)) {
```

Comment:
Oh, I forgot to suggest this earlier, but we could combine this into one call to setErrors, we don't need two. And we actually wouldn't have to account for asynchronous execution then, because we'd be setting the entire errors at once.

(That doesn't affect the other changes you made, just a side thing)